app/questionnaire/questionnaire_schema.py

```diff
@@ -91,6 +105,54 @@ def when_rules_section_dependencies_by_section(
     ) -> ImmutableDict[str, set[str]]:
         return ImmutableDict(self._when_rules_section_dependencies_by_section)
 
+    @cached_property
+    def when_rules_section_dependencies_for_progress(
+        self,
+    ) -> defaultdict[str, set[str]]:
+        """
+        This method flips the dependencies that were captured for progress value sources so that they can be
+        evaluated properly for when rules, this is because for when rules we need to check for dependencies in
+        previous sections, whereas for progress we are checking for dependent blocks/sections in "future" sections
+        """
+        when_rules_section_dependencies_for_progress = defaultdict(set)
+        for (
+            section,
+            dependent,
+        ) in (
+            self._when_rules_block_dependencies_by_section_for_progress_value_source.items()
+        ):
+            section_dependents = get_flattened_mapping_values(dependent)
+            for dependent_section in section_dependents:
+                when_rules_section_dependencies_for_progress[dependent_section].add(
+                    section
+                )
+
+        for (
+            section,
+            dependents,
+        ) in (
+            self._when_rules_section_dependencies_by_section_for_progress_value_source.items()
+        ):
+            for dependent_section in dependents:
+                when_rules_section_dependencies_for_progress[dependent_section].add(
+                    section
+                )
+        return when_rules_section_dependencies_for_progress
+
+    @cached_property
```

Comment:
I've ive misunderstood, please say, but I'm not sure I see the value in these @cached_property attributes, they're returning a cached reference to the object rather than a cached value, so if I did
`when_rules_section_dependencies_by_section_for_progress_value_source["x"] = OrderedSet(["y"])`
I would be updating `_when_rules_section_dependencies_by_section_for_progress_value_source` and updating the return value of the cached property. 

So I'm not sure they add any benefit over just making the attribute public, and I find them a bit confusing in this context

Reply:
For properties that don't do any operation but return an attribute, it could just be `@property` as caching would be wasteful. However, I would keep it as a property so it can't be updated by an external bit of code, if you made it public, then, any external interface can mutate it.

Let me know if I have misunderstood.

Reply:
because it returns a reference, would we not end up being able to get away with this sort of thing?
```
class X:

    def __init__(self) -> None:
        self._y: dict = {}

    @property
    def my_property(self):
        return self._y

x = X()
print(x.my_property)
x.my_property[1] = 2
print(x.my_property) # it updates
```

Reply:
Right I see, in that case we should keep it as cached_property but return it via `make_immutable` method we have. Thoughts?
We don't want the property or values to be mutated outside itself.

Reply:
This method wouldn't be called/updated this way, it is called and cached when we initialise the questionnaire schema class and then only checked when we are assessing the routing path.

Reply:
I'd be happy with that approach, removes the confusion and keeps its from being changed 👍

Reply:
> Will this be able to be made immutable since it is a defaultdict at the moment?

Ya why not?. It is a default dict when populating, we are just returning an immutable copy. It is still a dict that will be turned into ImmutableDict?.

Reply:
Using ImmutableDict now 👍